packages/nimble-components/src/table-column/number-text/models/format-helper.ts

```diff
@@ -0,0 +1,42 @@
+import type { TableColumnNumberTextColumnConfig } from '..';
+import type { TableNumberFieldValue } from '../../../table/types';
+
+export function formatNumber(
+    number: TableNumberFieldValue,
+    columnConfig: TableColumnNumberTextColumnConfig
+): string {
+    if (typeof number === 'number') {
+        try {
+            if (columnConfig.useDefaultFormatting) {
+                return formatNumberDefault(
+                    number,
+                    columnConfig.formatter,
+                    columnConfig.defaultScientificFormatter!
+                );
+            }
+            return columnConfig.formatter.format(number);
+        } catch (e) {
+            return '';
+        }
+    } else {
+        return '';
+    }
+}
+
+function formatNumberDefault(
+    number: number,
+    formatter: Intl.NumberFormat,
+    scientificFormatter: Intl.NumberFormat
+): string {
+    const absoluteValue = Math.abs(number);
+    if (absoluteValue > 0 && (absoluteValue < 1e-6 || absoluteValue >= 1e16)) {
+        // Intl.NumberFormat formats numbers like '1.23E4' and '1.23E-4' for scientific. We want the output to be closer
+        // to Number.toString(), i.e. lowercase 'e' and including the exponent + sign ('1.23e+4' and '1.23e-4').
+        const exponentSeparator = number > 1 ? 'e+' : 'e';
+        return scientificFormatter
+            .formatToParts(number)
+            .map(s => (s.type === 'exponentSeparator' ? exponentSeparator : s.value))
+            .join('');
+    }
+    return formatter.format(number);
```

Comment:
Where is this requirement coming from?

If it's to aid in testing or something then we should capture the logic in a page object instead. Similar to the [date formatter discussion](https://github.com/ni/nimble/pull/1363#discussion_r1268471402).

Otherwise I'd like to understand more why we aren't taking the formatting output as-is.

My concern is that this is delegating to an external library based on locale. I don't know that `e` is a world-wide globally accepted representation of exponentiation. That's what we are delegating to the browser for in the first place. I don't think we should presume to override it. I also don't think there is real significant value in overriding it.

Reply:
The [spec says 'formatting similarly to Number.toString()'](https://github.com/ni/nimble/blob/main/packages/nimble-components/src/table/specs/table-column-specs/table-column-formatted-text.md?plain=1#L174) which formats numbers for scientific notation differently than `Intl.NumberFormat` does by default. I also thought that the form with `e+` was easier to read (`1.22317846123e+21` vs `1.22317846123E21`, with the 2nd form the capital `E` almost blends into the numbers so it's harder to read to me / I prefer the 1st form) so I went with that for implementation.

It does complicate the code here though so it's worth discussion. I looked back at #887 and there wasn't really any discussion/examples for numbers with scientific notation there. We could always start with the default `Intl.NumberFormat` form and revisit if needed.